src/trace-plugin-interface.js

```diff
@@ -0,0 +1,242 @@
+'use strict';
+var cls = require('./cls.js');
+var constants = require('./constants.js');
+
+/**
+ * This file describes an interface for third-party plugins to enable tracing
+ * for arbitrary modules.
+ */
+
+/**
+ * An object that is associated with a single child span. It exposes
+ * functions for adding labels to or closing the associated span.
+ */
+function ChildSpan(agent, context) {
+  this.agent = agent;
+  this.context = context;
+}
+
+/**
+ * Adds a label to the underlying span.
+ * @param {string} key The name of the label to add.
+ * @param {*} value The value of the label to add.
+ */
+ChildSpan.prototype.addLabel = function(key, value) {
+  this.context.addLabel(key, value);
+};
+
+/**
+ * Ends the underlying span. This function should only be called once.
+ */
+ChildSpan.prototype.endSpan = function() {
+  this.context.close();
+};
+
+/**
+ * An object that is associated with a single root span. It exposes
+ * functions for adding labels to or closing the associated span.
+ */
+function Transaction(agent, context) {
+  this.agent = agent;
+  this.context = context;
+  this.closed = false;
+}
+
+/**
+ * Adds a label to the underlying span.
+ * @param {string} key The name of the label to add.
+ * @param {*} value The value of the label to add.
+ */
+Transaction.prototype.addLabel = function(key, value) {
+  this.context.addLabel(key, value);
+};
+
+/**
+ * Ends the underlying span. This function should only be called once.
+ */
+Transaction.prototype.endSpan = function() {
+  this.context.close();
+};
+
+/**
+ * Runs the given function in a child span, passing it an object that
+ * exposes an interface for adding labels and closing the span.
+ * @param {object} options An object that specifies options for how the child
+ * span is created and propogated.
+ * @param {string} options.name The name to apply to the child span.
+ * @param {?function(string, string)} options.setHeader A function describing
```

Comment:
I'm hoping for the latter; it's true that we only ever set one header field, but that header field name can only be exposed in one of two ways: either it's supplied as a constant (like api.TRACE_HEADER_NAME), or passed in as an argument to `(set)(get)Header`. So the two methods of setting the trace context header are:

```js
setTraceContextHeader(headerValue) {
  req.set(api.TRACE_HEADER_NAME, headerValue);
}
```

or

```js
setHeader(headerName, headerValue) {
  req.set(headerName, headerValue);
}
```

Between these I like the second because it seems like boilerplate to have to use a provided constant on an API in a function that is being provided to that same API. In addition, for servers with express-like request objects, you can pass in `req.set` and `req.get` directly for `(set)(get)Header` respectively, which is nice.